report/web.py

```diff
@@ -89,6 +93,115 @@ class Target:
   fixer_prompt: Optional[str] = None
 
 
+class FileSystem:
+  """
+  FileSystem provides a wrapper over standard library and GCS client and
+  automatically chooses which to use based on the provided path.
+  """
+
+  _gcs_client = None
+
+  @classmethod
+  def _get_gcs_client(cls):
+    """
+    Returns a cached storage client (a new one is created on first call.
+
+    A new client does authentication on first call, so caching the client will
+    same multiple authentication round trips to GCP.
+    """
+    if cls._gcs_client is None:
+      cls._gcs_client = storage.Client()
+
+    return cls._gcs_client
+
+  def __init__(self, path: str):
+    logging.debug('file operation %s', path)
+    self._path = path
+    self._gcs_bucket: Optional[storage.Bucket] = None
+
+    if path.startswith('gs://'):
+      path = path.removeprefix('gs://')
+      self._gcs_bucket = FileSystem._get_gcs_client().bucket(path.split('/')[0])
+      self._path = '/'.join(path.split('/')[1:])
+
+  def listdir(self) -> List[str]:
+    """listdir returns a list of files and directories in path."""
+    if self._gcs_bucket is not None:
+      # Make sure the path ends with a /, otherwise GCS just returns the
+      # directory as a prefix and not list the contents.
+      prefix = self._path
+      if not self._path.endswith('/'):
+        prefix = f'{self._path}/'
+
+      # Unfortunately GCS doesn't work like a normal file system and the client
+      # library doesn't even pretend there is a directory hierarchy.
+      # The list API does return a list of prefixes that we can join with the
+      # list of objects to get something close to listdir(). But client library
+      # is pretty weird and it stores the prefixes on the iterator...
+      # https://github.com/googleapis/python-storage/blob/64edbd922a605247203790a90f9536d54e3a705a/google/cloud/storage/client.py#L1356
+      it = self._gcs_bucket.list_blobs(prefix=prefix, delimiter='/')
+      paths = [f.name for f in it] + [p.removesuffix('/') for p in it.prefixes]
+      r = [p.removeprefix(prefix) for p in paths]
+      return r
+
+    return os.listdir(self._path)
+
+  def exists(self) -> bool:
+    """exists returns true if the path is a file or directory."""
+    if self._gcs_bucket is not None:
+      return self.isfile() or self.isdir()
+
+    return os.path.exists(self._path)
```

Comment:
Hmmm I'm not sure if `os.path.isfile() or os.path.isdir() == os.path.exists()` is guaranteed to be true in all file systems.

In GCS this holds because it's not a real filesystem.

Reply:
I think that's a valid assumption, but don't mind leaving it for now as you have merged.